lib/api/completeMultipartUpload.js

```diff
@@ -155,19 +157,71 @@ function completeMultipartUpload(authInfo, request, log, callback) {
+            // BACKWARD: Remove to remove the old splitter
+            if (mpuBucket.getMdBucketModelVersion() < 2) {
+                splitter = constants.oldSplitter;
+            }
+            // Reconstruct mpuOverviewKey to serve
+            // as key to pull metadata originally stored when mpu initiated
+            const mpuOverviewKey =
+                `overview${splitter}${objectKey}${splitter}${uploadId}`;
+
+            return metadata.getObjectMD(mpuBucket.getName(), mpuOverviewKey,
+            {}, log, (err, storedMetadata) => {
+                if (err) {
+                    return next(err, destBucket);
+                }
+                const location = storedMetadata.controllingLocationConstraint;
+                return next(null, destBucket, objMD, mpuBucket, jsonList,
+                storedMetadata, location, mpuOverviewKey);
+            });
+        },
+        function ifMultipleBackend(destBucket, objMD, mpuBucket, jsonList,
+        storedMetadata, location, mpuOverviewKey, next) {
+            if (config.backends.data === 'multiple') {
+                return multipleBackendGateway.completeMPU(objectKey,
+                uploadId, location, jsonList, destBucket, log,
+                (err, completeObjData) => {
+                    if (err) {
+                        return next(err, destBucket);
+                    }
+                    return next(null, destBucket, objMD, mpuBucket,
+                    jsonList, storedMetadata, completeObjData, mpuOverviewKey);
+                });
+            }
+            return next(null, destBucket, objMD, mpuBucket, jsonList,
+            storedMetadata, null, mpuOverviewKey);
+        },
+        function retrieveParts(destBucket, objMD, mpuBucket, jsonList,
+        storedMetadata, completeObjData, mpuOverviewKey, next) {
+            // means MPU completion was done on third-party backend,
+            // so part info is not in our metadata and related steps
+            // should be skipped
+            if (completeObjData) {
+                return next(null, destBucket, objMD, mpuBucket, null,
+                jsonList, storedMetadata, completeObjData, mpuOverviewKey);
+            }
+            return services.getMPUparts(mpuBucket.getName(), uploadId, log,
                 (err, result) => {
                     if (err) {
                         return next(err, destBucket);
                     }
                     const storedParts = result.Contents;
                     return next(null, destBucket, objMD, mpuBucket,
-                        storedParts, jsonList);
+                        storedParts, jsonList, storedMetadata, null,
+                        mpuOverviewKey);
                 });
         },
         function processParts(destBucket, objMD, mpuBucket, storedParts,
-            jsonList, next) {
+            jsonList, storedMetadata, completeObjData, mpuOverviewKey, next) {
+            let calculatedSize = 0;
```

Comment:
we shouldn't be storing a size of 0 for an mpu put on aws.  we will have to do a head request on the completed object and use that for the size.  we should add a test checking it.